tests/integration/abi_router_test.py

```diff
@@ -0,0 +1,540 @@
+        {DRProp.passed: True, DRProp.lastLog: "logging creation"},
+    ),
+    (
+        None,
+        pt.MethodConfig(
+            opt_in=pt.CallConfig.CALL,
+        ),
+        {
+            DRProp.passed: True,
+            DRProp.lastLog: lambda _, actual: actual
+            in (None, DryRunEncoder.hex("optin call")),
+        },
+    ),
+    (
+        CLEAR_STATE_CALL,
+        pt.MethodConfig(),  # ignored in this case
+        {
+            DRProp.passed: True,
+            DRProp.cost: 2,
+        },
+    ),
+]
+
+YACC_DRIVER = [case for case in QUESTIONABLE_DRIVER if case[0]]
+
+DRIVERS = {
+    "questionable": QUESTIONABLE_DRIVER,
+    "yacc": YACC_DRIVER,
+    "nontriv_clear": QUESTIONABLE_DRIVER,
+}
+
+
+def split_driver2predicates_methconfigs(driver) -> tuple[Predicates, ABICallConfigs]:
+    predicates = {}
+    methconfigs = {}
+    for meth, meth_config, predicate in driver:
+        predicates[meth] = predicate
+        if meth != CLEAR_STATE_CALL:
+            methconfigs[meth] = meth_config
+
+    return predicates, methconfigs
+
+
+def assert_full_method_coverage(router, methconfigs):
+    assert len(methconfigs) == len(rmc := router.method_configs)
+    for meth_sig, meth_config in rmc.items():
+        k = meth_sig
+        if k:
+            k = meth_sig.split("(")[0]
+        assert k in methconfigs, f"{k=} (derived from {meth_sig=} not in methconfigs"
+        assert meth_config == methconfigs[k]
+
+
+@pytest.mark.parametrize("case, version, router", ROUTER_CASES)
+def test_abi_router_positive(case, version, router):
+    """
+    Test the _positive_ version of a case. In other words, ensure that for each
+    router encountered and its driver, iterate through the driver as follows:
+        * consider each method or bare call
+        * consider each (OnComplete, CallConfig) combination
+        * assert that all predicates hold for this call
+    """
+    driver = DRIVERS[case]
+    predicates, methconfigs = split_driver2predicates_methconfigs(driver)
+
+    assert_full_method_coverage(router, methconfigs)
+
+    rsim = RouterSimulation(router, predicates)
+
+    def msg():
+        return f"""test_abi_router_positive()
+{case=}
+{version=}
+{router.name=}"""
+
+    results = rsim.simulate_and_assert(
+        approval_args_strat_type=RandomABIStrategyHalfSized,
+        clear_args_strat_type_or_inputs=RandomABIStrategy,
+        approval_abi_args_mod=None,
+        version=version,
+        method_configs=methconfigs,
+        num_dryruns=NUM_ROUTER_DRYRUNS,
+        msg=msg(),
+    )
+    # won't even get here if there was an error, but some extra sanity checks:
+    assert (sim_results := results.results) and all(
+        sim.succeeded for meth in sim_results.values() for sim in meth.values()
+    )
+
+    print("\nstats:", json.dumps(stats := results.stats, indent=2))
+    assert stats and all(stats.values())
+
+    # These fail because of differing scratch slot assignments:
```

Comment:
```suggestion
    # TODO: add these assertions after the flakiness of issue #199 is fixed for good
    # These fail because of differing scratch slot assignments:
```